### ISSUE TITLE
Server not receiving DATA in a Windows-Linux connection through a network.

### ISSUE THREAD

**Issue by delinage** [sctplab/usrsctp#573]
I am trying to connect two different computers through a network using usrsctp and I am having a really specific error. I am using the examples provided, so:

In the Windows computer I execute:
`echo_server.exe 9999 9899`

In the Linux computer (where X.X.X.X is the IP):
`./http_client  X.X.X.X 7 0 9899 9999`

The handshake is performed so there is connection from both endpoints but then, when the DATA packet is sent from the client, I get to receive it in the Windows computer (I attach the Wireshark capture) but it seems that the server is not receiving it. In fact, wireshark shows me a warning in the sequence numbers that might be the cause… (and the DATA is retransmitted)

If I use the same combination of examples in the same machine, e.g. Windows, I do get a DATA message and then a SACK message (that is what I expect in my Windows-Linux connection).

IMAGE: https://i.ibb.co/Y7c9kYs/Wireshark.png

What is the issue with the sequence numbers? it is fixable?

**Comment by tuexen** [comment]
Can you provide the `.pcapng` file instead of a screenshot? Are both physical machines in the same LAN?

**Comment by delinage** [comment]
I attach the wireshark capture (I replicated the error because I have lost previous capture, but it is the same).

Both machines are in the same network, meaning the do not have to go through internet NATs (and they should not).
 
https://wetransfer.com/downloads/0383779a4bbb0df5f5a58f930bbaf0f320210204110635/9654cc4b7002069375f3bcf20b0362e020210204110706/b7be17

**Comment by tuexen** [comment]
OK. From a client perspective, the tracefile looks very good. The handshake is successful, the DATA chunk is transmitted and retransmitted, because the peer does not send a SACK.

Where is the tracefile taken? At the client side or at the server side? Can you provide a tracefile taken at the server side?

Can you enable the the debug output on the server side? Compile with debug support and change in `echo_server.c` the line
```
	usrsctp_sysctl_set_sctp_debug_on(SCTP_DEBUG_NONE);
```
by
```
	usrsctp_sysctl_set_sctp_debug_on(SCTP_DEBUG_ALL);
```

Then provide the output.

**Comment by delinage** [comment]
That was from server side, let me attach the output you requested where everything is clearer in the file names:

https://wetransfer.com/downloads/12337fa645795b8f541c22ee6ce2bfbb20210204115615/073cab948748bfe64b766e652bad836620210204115645/73f213

While I was capturing everything I noticed an interesting thing: the hearbeats are being sent to an incorrect address. Let me explain you my addresses. I have windows (172.30.1.14) and linux (10.182.32.87) but in my windows I have more addresses that should not be in use as the 192.168.48.211 where the hearbeats are being sent to, might be related with the issue? I never told linux that address since I just write 172.30.1.14 in the client... but I see the server side share addresses in the INIT ACK so perhaps the client is not using them correctly? I am assuming since I am no expert in SCTP.

**Comment by tuexen** [comment]
The server announces in the INIT-ACK all the addresses configured at that host and the client is trying to send HEARTBEATs to all of them. That is expected.

However, I don't understand how your boxes can communicate at all. One side uses 10.182.32.87, the other 172.30.1.14. There must be a router in between... But that is not the problem.

The server sees the SCTP packets containing the `DATA` chunks:
```
[S][2.306] recv_function_udp: Received 92 bytes.[S][2.306]  - calling sctp_common_input_processing with off=12
[S][2.306] stcb:00000000 inp:01253490
[S][2.306] stcb is 00000000
[S][2.306] Ok, Common input processing called, m:01250D78 iphlen:0 offset:12 length:92 stcb:00000000
```
however, it does find the end-point (`inp:01253490`), but not the association (`stcb:00000000`).
I don't know why... Could you add `printf()'s` in `sctp_findassoc_by_vtag()` and possibly the functions it calls to figure out why it returns `NULL`? It should find the association.
It might be related to the addresses you use, but I don't see how right now.

**Comment by delinage** [comment]
So, I just throw some printf blindly as I show you in the attached txt ( https://wetransfer.com/downloads/87ed09ab58de20897852faaf5504e09520210204145136/c2407cbd314669b4c326df900025403520210204145159/10d3ba ) an this is the output (of the data retransmission):

```
> echo_server.exe 9999 9899
[sctp_findassoc_by_vtag] Entry point
[sctp_findassoc_by_vtag] SCTP_INP_INFO_RUNLOCK
Notification of length 25 received.
Notification of length 152 received.
[sctp_findassoc_by_vtag] Entry point
[sctp_findassoc_by_vtag] stcb->asoc.my_vtag == vtag
[sctp_findassoc_by_vtag] sctp_does_stcb_own_this_addr(stcb, to) == 0
[sctp_findassoc_by_vtag] address:172.30.1.14
[sctp_findassoc_by_vtag] SCTP_INP_INFO_RUNLOCK
[sctp_findassoc_by_vtag] Entry point
[sctp_findassoc_by_vtag] stcb->asoc.my_vtag == vtag
[sctp_findassoc_by_vtag] sctp_does_stcb_own_this_addr(stcb, to) == 0
[sctp_findassoc_by_vtag] address:172.30.1.14
[sctp_findassoc_by_vtag] SCTP_INP_INFO_RUNLOCK
[sctp_findassoc_by_vtag] Entry point
[sctp_findassoc_by_vtag] stcb->asoc.my_vtag == vtag
[sctp_findassoc_by_vtag] sctp_does_stcb_own_this_addr(stcb, to) == 0
[sctp_findassoc_by_vtag] address:172.30.1.14
[sctp_findassoc_by_vtag] SCTP_INP_INFO_RUNLOCK
[sctp_findassoc_by_vtag] Entry point
[sctp_findassoc_by_vtag] stcb->asoc.my_vtag == vtag
[sctp_findassoc_by_vtag] sctp_does_stcb_own_this_addr(stcb, to) == 0
[sctp_findassoc_by_vtag] address:172.30.1.14
[sctp_findassoc_by_vtag] SCTP_INP_INFO_RUNLOCK
```

Perhaps it thinks that 172.30.1.14 is not an accepted addr? Let me know if you need more printf in any specific location.

**Comment by tuexen** [comment]
Yeah, that is the conclusion. But why? Can you put in `printf()`s in `sctp_does_stcb_own_this_addr()` to figure out, why the address is not found. I guess we are in the `SCTP_PCB_FLAGS_BOUNDALL` branch...
